README.md
# msdf-bmfont-xml

[![Build Status](https://travis-ci.org/soimy/msdf-bmfont-xml.svg?branch=master)](https://travis-ci.org/soimy/msdf-bmfont-xml)
[![npm version](https://badge.fury.io/js/msdf-bmfont-xml.svg)](https://badge.fury.io/js/msdf-bmfont-xml)
![npm](https://img.shields.io/npm/dm/msdf-bmfont-xml.svg)

Converts a `.ttf` font file into multichannel signed distance fields, then outputs packed spritesheets and an `xml(.fnt}` / `txt(.fnt)` or `json` representation of an AngelCode BMFont file.

Signed distance fields are a method of reproducing vector shapes from a texture representation, popularized in [this paper by Valve](http://www.valvesoftware.com/publications/2007/SIGGRAPH2007_AlphaTestedMagnification.pdf).
This tool uses [Chlumsky/msdfgen](https://github.com/Chlumsky/msdfgen) to generate multichannel signed distance fields to preserve corners. The distance fields are created from vector fonts, then rendered into texture pages. A BMFont object is provided for character layout. (See: [BMFont format](http://www.angelcode.com/products/bmfont/doc/file_format.html))
![Preview image](https://raw.githubusercontent.com/soimy/msdf-bmfont-xml/master/assets/msdf-bmfont-xml.png)

## Run script to see MSDF font in browser using pixi.js

```bash
npm install & npm run render
```

## Install as CLI

```bash
npm install msdf-bmfont-xml -g
```

Then you just need to call `msdf-bmfont` from console to generate font file.
Type in `msdf-bmfont --help` for more detail usage.

![Console-Demo](https://raw.githubusercontent.com/soimy/msdf-bmfont-xml/master/assets/console-demo.gif)

### Usage

```bash
Usage: msdf-bmfont [options] <font-file>

Creates a BMFont compatible bitmap font of signed distance fields from a font file

Options:
  -V, --version                 output the version number
  -f, --output-type <format>    font file format: xml(default) | json | txt (default: "xml")
  -o, --filename <atlas_path>   filename of font textures (defaut: font-face)
                                font filename always set to font-face name
  -s, --font-size <fontSize>    font size for generated textures (default: 42)
  -i, --charset-file <charset>  user-specified charactors from text-file
  -m, --texture-size <w,h>      ouput texture atlas size (default: [2048,2048])
  -p, --texture-padding <n>     padding between glyphs (default: 1)
  -b, --border <n>              space between glyphs textures & edge (default: 0)
  -r, --distance-range <n>      distance range for SDF (default: 4)
  -t, --field-type <type>       msdf(default) | sdf | psdf (default: "msdf")
  -d, --round-decimal <digit>   rounded digits of the output font file. (default: 0)
  -v, --vector                  generate svg vector file for debuging
  -u, --reuse [file.cfg]        save/create config file for reusing settings (default: false)
      --smart-size              shrink atlas to the smallest possible square
      --pot                     atlas size shall be power of 2
      --square                  atlas size shall be square
      --rot                     allow 90-degree rotation while packing
      --rtl                     use RTL(Arabic/Persian) charactors fix
  -h, --help                    output usage information
```

### CLI Examples

Generate a multi-channel signed distance field font atlas with ASCII charset, font size 42, spread 3, maximum texture size 512x256, padding 1, and save out config file:

```bash
msdf-bmfont --reuse -o path/to/atlas.png -m 512,256 -s 42 -r 3 -p 1 -t msdf path/to/font.ttf
```

We will get three file: `atlas.0.png` `atlas.0.cfg` & `font.fnt` and this is the generated atlas in the minimum pot size (256x256):

![Atlas0](https://raw.githubusercontent.com/soimy/msdf-bmfont-xml/master/assets/atlas.0.png)

Then we want to use the old setting except a different font and use monochrome signed distance field atlas, and output an extra `.svg` version of atlas:

```bash
msdf-bmfont -v -u path/to/atlas.0.cfg -t sdf -p 0 -r 8 path/to/anotherfont.ttf
```

This time we get a modified `atlas.0.png` with new bitmap font appended:

![Atlas1](https://raw.githubusercontent.com/soimy/msdf-bmfont-xml/master/assets/atlas.1.jpg)

Not satisfied with the style? Remember we got a `svg` atlas!

![svg](https://raw.githubusercontent.com/soimy/msdf-bmfont-xml/master/assets/svg.png)

How about fire up some graphic editor and add some neat effect and lay on the output atlas?

![final](https://raw.githubusercontent.com/soimy/msdf-bmfont-xml/master/assets/atlas.2.jpg)

## Install as Module

```bash
npm install msdf-bmfont-xml
```

### Module usage Examples

Writing the distance fields and font data to disk:

```js
const generateBMFont = require('msdf-bmfont-xml');
const fs = require('fs');

generateBMFont('Some-Font.ttf', (error, textures, font) => {
  if (error) throw error;
  textures.forEach((texture, index) => {
    fs.writeFile(texture.filename, texture.texture, (err) => {
      if (err) throw err;
    });
  });
  fs.writeFile(font.filename, font.data, (err) => {
    if (err) throw err;
  });
});
```

Generating a single channel signed distance field with a custom character set:

```js
const generateBMFont = require('msdf-bmfont');

const opt = {
  charset: 'ABC.ez_as-123!',
  fieldType: 'sdf'
};
generateBMFont('Some-Font.ttf', opt, (error, textures, font) => {
  ...
});
```

### API

#### `generateBMFont(fontPath | fontBuffer, [opt], callback)`

Renders a bitmap font from the font specified by `fontPath` or `fontBuffer`, with optional `opt` settings, triggering `callback` on complete.

Options:

- `outputType` (String)
  - type of output font file. Defaults to `xml`
    - `xml` a BMFont standard .fnt file which is wildly supported.
    - `json` a JSON file compatible with [Hiero](https://github.com/libgdx/libgdx/wiki/Hiero)
- `filename` (String)
  - filename of both font file and font atlas. If omited, font face name is used. **Required** if font is provided as a Buffer.
- `charset` (String|Array)
  - the characters to include in the bitmap font. Defaults to all ASCII printable characters.
- `fontSize` (Number)
  - the font size at which to generate the distance field. Defaults to `42`
- `textureSize` (Array[2])
  - the dimensions of an output texture sheet, normally power-of-2 for GPU usage. Both dimensions default to `[512, 512]`
- `texturePadding` (Number)
  - pixels between each glyph in the texture. Defaults to `2`
- `border` (Number)
  - space between glyphs textures & edge. Defaults to `0`
- `fieldType` (String)
  - what kind of distance field to generate. Defaults to `msdf`. Must be one of:
    - `msdf` Multi-channel signed distance field
    - `sdf` Monochrome signed distance field
    - `psdf` monochrome signed pseudo-distance field
- `distanceRange` (Number)
  - the width of the range around the shape between the minimum and maximum representable signed distance in pixels, defaults to `3`
- `roundDecimal` (Number)
  - rounded digits of the output font metics. For `xml` output, `roundDecimal: 0` recommended.
- `vector` (Boolean)
  - output a SVG Vector file for debugging. Defauts to `false`
- `smart-size` (Boolean)
  - shrink atlas to the smallest possible square. Default: `false`
- `pot` (Boolean)
  - output atlas size shall be power of 2. Default: `false`
- `square` (Boolean)
  - output atlas size shall be square. Default: `false`
- `rot` (Boolean)
  - allow 90-degree rotation while packing. Default: `false`
- `rtl` (Boolean)
  - use RTL(Arabic/Persian) charators fix. Default: `false`

The `callback` is called with the arguments `(error, textures, font)`

- `error` on success will be null/undefined
- `textures` an array of js objects of texture spritesheet.
  - `textures[index].filename` Spritesheet filename
  - `textures[index].texture` Image Buffers, containing the PNG data of one texture sheet
- `font` an object containing the BMFont data, to be used to render the font
  - `font.filename` font filename
  - `font.data` stringified xml\json data to be written to disk

Since `opt` is optional, you can specify `callback` as the second argument.

## License

MIT, see [LICENSE.md](http://github.com/Jam3/xhr-request/blob/master/LICENSE.md) for details.


CHANGELOG.md
# Changelog

All notable changes to this project will be documented in this file. See [standard-version](https://github.com/conventional-changelog/standard-version) for commit guidelines.

## [2.7.0](https://github.com/soimy/msdf-bmfont-xml/compare/v2.6.0...v2.7.0) (2022-07-23)


### Features

* add new output type: txt ([34d8976](https://github.com/soimy/msdf-bmfont-xml/commit/34d8976a8672045534c8c379c967cf16bcbb8cbe))

## [2.6.0](https://github.com/soimy/msdf-bmfont-xml/compare/v2.5.4...v2.6.0) (2022-01-31)


### Features

* Add version update-notifier ([23aace9](https://github.com/soimy/msdf-bmfont-xml/commit/23aace97c8e4af5174f8191133e42b54baafd0b5))

### [2.5.4](https://github.com/soimy/msdf-bmfont-xml/compare/v2.5.3...v2.5.4) (2020-02-08)


### Bug Fixes

* CLI&document for SVG debug generation from `-t svg` to `-v` ([c37cc40](https://github.com/soimy/msdf-bmfont-xml/commit/c37cc40)), closes [#46](https://github.com/soimy/msdf-bmfont-xml/issues/46)
* omit path in font-face ([ae559fd](https://github.com/soimy/msdf-bmfont-xml/commit/ae559fd)), closes [#49](https://github.com/soimy/msdf-bmfont-xml/issues/49)
* Re-use cfg settings not applied ([3b57562](https://github.com/soimy/msdf-bmfont-xml/commit/3b57562)), closes [#44](https://github.com/soimy/msdf-bmfont-xml/issues/44)



### [2.5.3](https://github.com/soimy/msdf-bmfont-xml/compare/v2.5.2...v2.5.3) (2020-01-26)



### [2.5.2](https://github.com/soimy/msdf-bmfont-xml/compare/v2.5.1...v2.5.2) (2019-07-12)


### Bug Fixes

* async texture generation skips fontData stringify ([230d0b3](https://github.com/soimy/msdf-bmfont-xml/commit/230d0b3)), closes [#39](https://github.com/soimy/msdf-bmfont-xml/issues/39)



### [2.5.1](https://github.com/soimy/msdf-bmfont-xml/compare/v2.5.0...v2.5.1) (2019-07-11)


### Bug Fixes

* async jimp.read() should be fulfilled before continue packing ([5d95b32](https://github.com/soimy/msdf-bmfont-xml/commit/5d95b32))
* package script render changed to yarn ([84732bf](https://github.com/soimy/msdf-bmfont-xml/commit/84732bf))


### Tests

* Fix run render by replacing abandoned devtools ([071b016](https://github.com/soimy/msdf-bmfont-xml/commit/071b016)), closes [#38](https://github.com/soimy/msdf-bmfont-xml/issues/38)



## [2.5.0](https://github.com/soimy/msdf-bmfont-xml/compare/v2.4.3...v2.5.0) (2019-06-19)


### Features

* new border option to control packing space to atlas edge ([33598b9](https://github.com/soimy/msdf-bmfont-xml/commit/33598b9))


index.js
const utils = require('./lib/utils');
const reshaper = require('arabic-persian-reshaper').ArabicShaper;
const opentype = require('opentype.js');
const exec = require('child_process').exec;
const mapLimit = require('map-limit');
const MaxRectsPacker = require('maxrects-packer').MaxRectsPacker;
const path = require('path');
const ProgressBar = require('cli-progress');
const fs = require('fs');
const buffer = require('buffer').Buffer;
const Jimp = require('jimp');
const readline = require('readline');
const assert = require('assert');

const defaultCharset = " !\"#$%&'()*+,-./0123456789:;<=>?@ABCDEFGHIJKLMNOPQRSTUVWXYZ[\\]^_`abcdefghijklmnopqrstuvwxyz{|}~".split('');
const controlChars = ['\n', '\r', '\t'];

const binaryLookup = {
  darwin: 'msdfgen.osx',
  win32: 'msdfgen.exe',
  linux: 'msdfgen.linux'
};

module.exports = generateBMFont;

function getBinaryDir(){
  switch(process.platform) {
      case 'win32':
          return process.arch === 'x64' ? 'win64' : 'win32';
      case 'linux':
          return process.arch === 'arm64' ? 'linux_arm64' : 'linux';
      case 'darwin':
          return 'darwin';
  }
}

/**
 * Creates a BMFont compatible bitmap font of signed distance fields from a font file
 *
 * @param {string|Buffer} fontPath - Path or Buffer for the input ttf/otf/woff font
 * @param {Object} opt - Options object for generating bitmap font (Optional) :
 *            outputType : font file format Avaliable: xml(default), json, txt
 *            filename : filename of both font file and font textures
 *            fontSize : font size for generated textures (default 42)
 *            charset : charset in generated font, could be array or string (default is Western)
 *            textureWidth : Width of generated textures (default 512)
 *            textureHeight : Height of generated textures (default 512)
 *            distanceRange : distance range for computing signed distance field
 *            fieldType : "msdf"(default), "sdf", "psdf"
 *            roundDecimal  : rounded digits of the output font file. (Defaut is null)
 *            smartSize : shrink atlas to the smallest possible square (Default: false)
 *            pot : atlas size shall be power of 2 (Default: false)
 *            square : atlas size shall be square (Default: false)
 *            rot : allow 90-degree rotation while packing (Default: false)
 *            rtl : use RTL charators fix (Default: false)
 * @param {function(string, Array.<Object>, Object)} callback - Callback funtion(err, textures, font)
 *
 */
function generateBMFont (fontPath, opt, callback) {
  if (typeof opt === 'function') {
    callback = opt;
    opt = {};
  }

  const binName = binaryLookup[process.platform];
  const binDir = getBinaryDir();

  assert.ok(binName, `No msdfgen binary for platform ${process.platform}.`);
  assert.ok(fontPath, 'must specify a font path');
  assert.ok(typeof fontPath === 'string' || fontPath instanceof Buffer, 'font must be string path or Buffer');
  assert.ok(opt.filename || !(fontPath instanceof Buffer), 'must specify filename if font is a Buffer');
  assert.ok(callback, 'missing callback')
  assert.ok(typeof callback === 'function', 'expected callback to be a function');
  assert.ok(!opt.textureSize || opt.textureSize.length === 2, 'textureSize format shall be: width,height');

  // Set fallback output path to font path
  let fontDir = typeof fontPath === 'string' ? path.dirname(fontPath) : '';
  const binaryPath = path.join(__dirname, 'bin', binDir, binName);

  // const reuse = (typeof opt.reuse === 'boolean' || typeof opt.reuse === 'undefined') ? {} : opt.reuse.opt;
  let reuse, cfg = {};
  if (typeof opt.reuse !== 'undefined' && typeof opt.reuse !== 'boolean') {
    if (!fs.existsSync(opt.reuse)) {
      console.log('Creating cfg file : ' + opt.reuse);
      reuse = {};
    } else {
      console.log('Loading cfg file : ' + opt.reuse);
      cfg = JSON.parse(fs.readFileSync(opt.reuse, 'utf8'));
      reuse = cfg.opt;
    }
  } else reuse = {};
  const outputType = opt.outputType = utils.valueQueue([opt.outputType, reuse.outputType, "xml"]);
  let filename = utils.valueQueue([opt.filename, reuse.filename]);
  const distanceRange = opt.distanceRange = utils.valueQueue([opt.distanceRange, reuse.distanceRange, 4]);
  const fontSize = opt.fontSize = utils.valueQueue([opt.fontSize, reuse.fontSize, 42]);
  const fontSpacing = opt.fontSpacing = utils.valueQueue([opt.fontSpacing, reuse.fontSpacing, [0, 0]]);
  const pad = distanceRange >> 1;
  const fontPadding = opt.fontPadding = utils.valueQueue([opt.fontPadding, reuse.fontPadding, [pad, pad, pad, pad]]);
  const textureWidth = opt.textureWidth = utils.valueQueue([opt.textureSize || reuse.textureSize, [512, 512]])[0];
  const textureHeight = opt.textureHeight = utils.valueQueue([opt.textureSize || reuse.textureSize, [512, 512]])[1];
  const texturePadding = opt.texturePadding = utils.valueQueue([opt.texturePadding, reuse.texturePadding, 1]);
  const border = opt.border = utils.valueQueue([opt.border, reuse.border, 0]);
  const fieldType = opt.fieldType = utils.valueQueue([opt.fieldType, reuse.fieldType, 'msdf']);
  const roundDecimal = opt.roundDecimal = utils.valueQueue([opt.roundDecimal, reuse.roundDecimal]); // if no roudDecimal option, left null as-is
  const smartSize = opt.smartSize = utils.valueQueue([opt.smartSize, reuse.smartSize, false]);
  const pot = opt.pot = utils.valueQueue([opt.pot, reuse.pot, false]);
  const square = opt.square = utils.valueQueue([opt.square, reuse.square, false]);
  const debug = opt.vector || false;
  const tolerance = opt.tolerance = utils.valueQueue([opt.tolerance, reuse.tolerance, 0]);
  const isRTL = opt.rtl = utils.valueQueue([opt.rtl, reuse.rtl, false]);
  const allowRotation = opt.rot = utils.valueQueue([opt.rot, reuse.rot, false]);
  if (isRTL) opt.charset = reshaper.convertArabic(opt.charset);
  let charset = opt.charset = (typeof opt.charset === 'string' ? Array.from(opt.charset) : opt.charset) || reuse.charset || defaultCharset;

  // TODO: Validate options
  if (fieldType !== 'msdf' && fieldType !== 'sdf' && fieldType !== 'psdf') {
    throw new TypeError('fieldType must be one of msdf, sdf, or psdf');
  }

  const font = typeof fontPath === 'string'
    ? opentype.loadSync(fontPath)
    : opentype.parse(utils.bufferToArrayBuffer(fontPath));

  if (font.outlinesFormat !== 'truetype' && font.outlinesFormat !== 'cff') {
    throw new TypeError('must specify a truetype font (ttf, otf, woff)');
  }
  const packer = new MaxRectsPacker(textureWidth, textureHeight, texturePadding, {
    smart: smartSize,
    pot: pot,
    square: square,
    allowRotation: allowRotation,
    tag: false,
    border: border
  });
  const chars = [];

  charset = charset.filter((e, i, self) => {
    return (i == self.indexOf(e)) && (!controlChars.includes(e));
  }); // Remove duplicate & control chars

  const os2 = font.tables.os2;
  const baseline = os2.sTypoAscender * (fontSize / font.unitsPerEm) + (distanceRange >> 1);

  const fontface = typeof fontPath === 'string' ? path.basename(fontPath, path.extname(fontPath)) : filename;

  if(!filename) {
    filename = fontface;
    console.log(`Use font-face as filename : ${filename}`);
  } else {
    if (opt.filename) fontDir = path.dirname(opt.filename);
    filename = opt.filename = path.basename(filename, path.extname(filename));
  }

  // Initialize settings
  let settings = {};
  settings.opt = JSON.parse(JSON.stringify(opt));
  delete settings.opt['reuse']; // prune previous settings
  let pages = [];
  if (cfg.packer !== undefined) {
    pages = cfg.pages;
    packer.load(cfg.packer.bins);
  }

  let bar;
  bar = new ProgressBar.Bar({
    format: "Generating {percentage}%|{bar}| ({value}/{total}) {duration}s",
    clearOnComplete: true
  }, ProgressBar.Presets.shades_classic); 
  bar.start(charset.length, 0);

  mapLimit(charset, 15, (char, cb) => {
    generateImage({
      binaryPath,
      font,
      char,
      fontSize,
      fieldType,
      distanceRange,
      roundDecimal,
      debug,
      tolerance
    }, (err, res) => {
      if (err) return cb(err);
      bar.increment();
      cb(null, res);
    });
  }, async (err, results) => {
    if (err) callback(err);
    bar.stop();

    packer.addArray(results);
    const textures = packer.bins.map(async (bin, index) => {
      let svg = "";
      let texname = "";
      let fillColor = fieldType === "msdf" ? 0x000000ff : 0x00000000;
      let img = new Jimp(bin.width, bin.height, fillColor);
      if (index > pages.length - 1) { 
        if (packer.bins.length > 1) texname = `${filename}.${index}`;
        else texname = filename; 
        pages.push(`${texname}.png`);
      } else {
        texname = path.basename(pages[index], path.extname(pages[index]));
        let imgPath = path.join(fontDir, `${texname}.png`);
        // let imgPath = `${texname}.png`;
        console.log('Loading previous image : ', imgPath);
        const loader = Jimp.read(imgPath);
        loader.catch(err => {
          console.warn("File read error: ", err);
        });
        const prevImg = await loader;
        img.composite(prevImg, 0, 0);
      }
      bin.rects.forEach(rect => {
        if (rect.data.imageData) {
          if (rect.rot) {
            rect.data.imageData.rotate(90);
          }
          img.composite(rect.data.imageData, rect.x, rect.y);
          if (debug) {
            const x_woffset = rect.x - rect.data.fontData.xoffset + (distanceRange >> 1);
            const y_woffset = rect.y - rect.data.fontData.yoffset + baseline + (distanceRange >> 1);
            svg += font.charToGlyph(rect.data.fontData.char).getPath(x_woffset, y_woffset, fontSize).toSVG() + "\n";
          }
        }
        const charData = rect.data.fontData;
        charData.x = rect.x;
        charData.y = rect.y;
        charData.page = index;
        chars.push(rect.data.fontData);
      });
      const buffer = await img.getBufferAsync(Jimp.MIME_PNG);
      let tex = {
        filename: path.join(fontDir, texname),
        texture: buffer 
      }
      if (debug) tex.svg = svg;
      return tex;
    });

    const asyncTextures = await Promise.all(textures);

    const kernings = [];
    charset.forEach(first => {
      charset.forEach(second => {
        const amount = font.getKerningValue(font.charToGlyph(first), font.charToGlyph(second));
        if (amount !== 0) {
          kernings.push({
            first: first.charCodeAt(0),
            second: second.charCodeAt(0),
            amount: amount * (fontSize / font.unitsPerEm)
          });
        }
      });
    });

    const fontData = {
      pages,
      chars,
      info: {
        face: fontface,
        size: fontSize,
        bold: 0,
        italic: 0,
        charset,
        unicode: 1,
        stretchH: 100,
        smooth: 1,
        aa: 1,
        padding: fontPadding,
        spacing: fontSpacing,
        outline: 0
      },
      common: {
        lineHeight: (os2.sTypoAscender - os2.sTypoDescender + os2.sTypoLineGap) * (fontSize / font.unitsPerEm),
        base: baseline,
        scaleW: packer.bins[0].width,
        scaleH: packer.bins[0].height,
        pages: packer.bins.length,
        packed: 0,
        alphaChnl: 0,
        redChnl: 0,
        greenChnl: 0,
        blueChnl: 0
      },
      distanceField: {
        fieldType: fieldType,
        distanceRange: distanceRange
      },
      kernings: kernings
    };
    if(roundDecimal !== null) utils.roundAllValue(fontData, roundDecimal, true);
    let fontFile = {};
    const ext = outputType === "json" ? `.json` : `.fnt`;
    fontFile.filename = path.join(fontDir, fontface + ext);
    fontFile.data = utils.stringify(fontData, outputType);

    // Store pages name and available packer freeRects in settings
    settings.pages = pages;
    settings.packer = {};
    settings.packer.bins = packer.save(); 
    fontFile.settings = settings;

    console.log("\nGeneration complete!\n");
    callback(null, asyncTextures, fontFile);
  });
}

function generateImage (opt, callback) {
  const {binaryPath, font, char, fontSize, fieldType, distanceRange, roundDecimal, debug, tolerance} = opt;
  const glyph = font.charToGlyph(char);
  const commands = glyph.getPath(0, 0, fontSize).commands;
  let contours = [];
  let currentContour = [];
  const bBox = glyph.getPath(0, 0, fontSize).getBoundingBox();
  commands.forEach(command => {
    if (command.type === 'M') { // new contour
      if (currentContour.length > 0) {
        contours.push(currentContour);
        currentContour = [];
      }
    }
    currentContour.push(command);
  });
  contours.push(currentContour);

  if (tolerance != 0) {
    utils.setTolerance(tolerance, tolerance * 10);
    let numFiltered = utils.filterContours(contours);
    if (numFiltered && debug)
      console.log(`\n${char} removed ${numFiltered} small contour(s)`);
    // let numReversed = utils.alignClockwise(contours, false);
    // if (numReversed && debug)
    //   console.log(`${char} found ${numReversed}/${contours.length} reversed contour(s)`);
  }
  let shapeDesc = utils.stringifyContours(contours);

  if (contours.some(cont => cont.length === 1)) console.log('length is 1, failed to normalize glyph');
  const scale = fontSize / font.unitsPerEm;
  const baseline = font.tables.os2.sTypoAscender * (fontSize / font.unitsPerEm);
  const pad = distanceRange >> 1;
  let width = Math.round(bBox.x2 - bBox.x1) + pad + pad;
  let height = Math.round(bBox.y2 - bBox.y1) + pad + pad;
  let xOffset = Math.round(-bBox.x1) + pad;
  let yOffset = Math.round(-bBox.y1) + pad;
  if (roundDecimal != null) {
    xOffset = utils.roundNumber(xOffset, roundDecimal);
    yOffset = utils.roundNumber(yOffset, roundDecimal);
  }
  let command = `"${binaryPath}" ${fieldType} -format text -stdout -size ${width} ${height} -translate ${xOffset} ${yOffset} -pxrange ${distanceRange} -stdin`;

  let subproc = exec(command, (err, stdout, stderr) => {
    if (err) return callback(err);
    const rawImageData = stdout.match(/([0-9a-fA-F]+)/g).map(str => parseInt(str, 16)); // split on every number, parse from hex
    const pixels = [];
    const channelCount = rawImageData.length / width / height;

    if (!isNaN(channelCount) && channelCount % 1 !== 0) {
      console.error(command);
      console.error(stdout);
      return callback(new RangeError('msdfgen returned an image with an invalid length'));
    }
    if (fieldType === 'msdf') {
      for (let i = 0; i < rawImageData.length; i += channelCount) {
        pixels.push(...rawImageData.slice(i, i + channelCount), 255); // add 255 as alpha every 3 elements
      }
    } else {
      for (let i = 0; i < rawImageData.length; i += channelCount) {
        pixels.push(rawImageData[i], rawImageData[i], rawImageData[i], rawImageData[i]); // make monochrome w/ alpha
      }
    }
    let imageData;
    if (isNaN(channelCount) || !rawImageData.some(x => x !== 0)) { // if character is blank
      readline.clearLine(process.stdout);
      readline.cursorTo(process.stdout, 0);
      console.log(`Warning: no bitmap for character '${char}' (${char.charCodeAt(0)}), adding to font as empty`);
      width = 0;
      height = 0;
    } else {
      const buffer = new Uint8ClampedArray(pixels);
      imageData = new Jimp({data: buffer, width: width, height: height});
    }
    const container = {
      data: {
        imageData,
        fontData: {
          id: char.charCodeAt(0),
          index: glyph.index,
          char: String(char),
          width: width,
          height: height,
          xoffset: Math.round(bBox.x1) - pad,
          yoffset: Math.round(bBox.y1) + pad + baseline,
          xadvance: glyph.advanceWidth * scale,
          chnl: 15
        }
      },
      width: width,
      height: height
    };
    callback(null, container);
  });
  
  subproc.stdin.write(shapeDesc);
  subproc.stdin.write('\n');
  subproc.stdin.destroy();
}



cli.js
#!/usr/bin/env node

const pkg = require('./package.json');
const generateBMFont = require('./index');
const fs = require('fs');
const path = require('path');
const handlebars = require('handlebars');
const args = require('commander');
const updateNotifier = require('update-notifier');
const utils = require('./lib/utils');

updateNotifier({pkg}).notify();

let fontFile;
args
  .version('msdf-bmfont-xml v' + pkg.version)
  .usage('[options] <font-file>')
  .arguments('<font_file>')
  .description('Creates a BMFont compatible bitmap font of signed distance fields from a font file')
  .option('-f, --output-type <format>', 'font file format: xml(default) | json | txt', /^(xml|json|txt)$/i, 'xml')
  .option('-o, --filename <atlas_path>', 'filename of font textures (defaut: font-face) font filename always set to font-face name')
  .option('-s, --font-size <fontSize>', 'font size for generated textures', 42)
  .option('-i, --charset-file <charset>', 'user-specified charactors from text-file', fileExistValidate)
  .option('-m, --texture-size <w,h>', 'ouput texture atlas size', (v) => {return v.split(',')}, [2048, 2048])
  .option('-p, --texture-padding <n>', 'padding between glyphs', 1)
  .option('-b, --border <n>', 'space between glyphs textures & edge', 0)
  .option('-r, --distance-range <n>', 'distance range for SDF', 4)
  .option('-t, --field-type <type>', 'msdf(default) | sdf | psdf', /^(msdf|sdf|psdf)$/i, 'msdf')
  .option('-d, --round-decimal <digit>', 'rounded digits of the output font file.', 0)
  .option('-v, --vector', 'generate svg vector file for debuging', false)
  .option('-u, --reuse [file.cfg]', 'save/create config file for reusing settings', false)
  .option('    --smart-size', 'shrink atlas to the smallest possible square', true)
  .option('    --pot', 'atlas size shall be power of 2', false)
  .option('    --square', 'atlas size shall be square', false)
  .option('    --rot', 'allow 90-degree rotation while packing', false)
  .option('    --rtl', 'use RTL(Arabic/Persian) charactors fix', false)
  .action(function(file){
    fontFile = fileExistValidate(file);
  }).parse(process.argv);

//
// Initialize options 
//
let opt = args.opts();
utils.roundAllValue(opt); // Parse all number from string
if (!fontFile) {
  console.error('Must specify font-file, use: \'msdf-bmfont -h\' for more infomation');
  process.exit(1);
}
const fontface = path.basename(fontFile, path.extname(fontFile));
const fontDir = path.dirname(fontFile);

//
// Set default value
//
// Note: somehow commander.js didn't parse boolean default value
// need to feed manually
//
opt.fontFile = fontFile;
if (typeof opt.reuse === 'boolean') {
  opt.filename = utils.valueQueue([opt.filename, path.join(fontDir, fontface)]);
  opt.vector = utils.valueQueue([opt.vector, false]);
  opt.reuse = utils.valueQueue([opt.reuse, false]);
  opt.smartSize = utils.valueQueue([opt.smartSize, false]);
  opt.pot = utils.valueQueue([opt.pot, false]);
  opt.square = utils.valueQueue([opt.square, false]);
  opt.rot = utils.valueQueue([opt.rot, false]);
  opt.rtl = utils.valueQueue([opt.rtl, false]);
} else {
  opt.filename = utils.valueQueue([opt.filename, path.join("./", fontface)]);
}

//
// Display options 
//
const keys = Object.keys(opt)
const padding = longestLength(keys) + 2;
console.log("\nUsing following settings");
console.log("========================================");
keys.forEach(key => {
  if (typeof opt.reuse === 'string' && typeof opt[key] === 'undefined') {
    console.log(pad(key, padding) + ": Defined in [" + opt.reuse + "]");
  } else if (key === 'charsetFile' && typeof opt[key] === 'undefined') {
    console.log(pad(key, padding) + ": Unspecified, fallback to ASC-II");
  } else console.log(pad(key, padding) + ": " + opt[key]);
});
console.log("========================================");

//
// Validate
//
if (typeof opt.fontFile === 'undefined') {
  console.error('No font file specified, aborting.... use -h for help');
  process.exit(1);
}
if (typeof opt.reuse !== 'boolean') opt.reuse = fileValidate(opt.reuse);

fs.readFile(opt.charsetFile || '', 'utf8', (error, data) => {
  if (error) {
    console.warn('No valid charset file loaded, fallback to ASC-II');
  }
  if (data) opt.charset = data;
  
  generateBMFont(opt.fontFile, opt, (error, textures, font) => {
    if (error) throw error;
    textures.forEach((texture, index) => {
      if (opt.vector) {
        const svgTemplate = 
        `<?xml version="1.0"?>
<!DOCTYPE svg PUBLIC "-//W3C//DTD SVG 1.1//EN" "http://www.w3.org/Graphics/SVG/1.1/DTD/svg11.dtd">
<svg xmlns="http://www.w3.org/2000/svg" version="1.1" width="{{width}}" height="{{height}}">
{{{svgPath}}}
</svg>`;
        const template = handlebars.compile(svgTemplate);
        const content = template({
          width: opt.textureWidth,
          height: opt.textureHeight,
          svgPath: texture.svg
        });
        fs.writeFile(`${texture.filename}.svg`, content , (err) => {
          if (err) throw err;
          console.log('wrote svg[', index, ']         : ', `${texture.filename}.svg`);
        });
      } 
      fs.writeFile(`${texture.filename}.png`, texture.texture, (err) => {
        if (err) throw err;
        console.log('wrote spritesheet[', index, '] : ', `${texture.filename}.png`);
      });
    });
    fs.writeFile(font.filename, font.data, (err) => {
      if (err) throw err;
      console.log('wrote font file        : ', font.filename);
    });
    if(opt.reuse !== false) {
      let cfgFileName = typeof opt.reuse === 'boolean' ? `${textures[0].filename}.cfg` : opt.reuse;
      fs.writeFile(cfgFileName, JSON.stringify(font.settings, null, '\t'), (err) => {
        if (err) throw err;
        console.log('wrote cfg file         : ', cfgFileName);
      });
    }
  });
});

/**
 * Pad `str` to `width`.
 *
 * @param {String} str
 * @param {Number} width
 * @return {String}
 * @api private
 */
function pad(str, width) {
  var len = Math.max(0, width - str.length);
  return str + Array(len + 1).join(' ');
}

/**
 * Return the largest length of string array.
 *
 * @param {Array.<String>} arr
 * @return {Number}
 * @api private
 */

function longestLength(arr) {
  return arr.reduce((max, element) => {
    return Math.max(max, element.length);
  }, 0);
};

function fileExistValidate(filePath) {
  try {
    if(fs.statSync(filePath).isFile()) return path.normalize(filePath);
    else {
      console.error('File: ', filePath, ' not found! Aborting...');
      process.exit(1);
    }
  } catch(err) {
      console.error('File: ', filePath, ' not valid! Aborting...');
      process.exit(1);
  }
}

function fileValidate(filePath) {
  if (require('is-invalid-path')(filePath)) {
    console.error('File: ', filePath, ' not valid! Aborting...');
    process.exit(1);
  } else return path.normalize(filePath);
}